src/DataStructures/SparseMatrixFiller.cpp

```diff
@@ -0,0 +1,109 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "DataStructures/SparseMatrixFiller.hpp"
+
+#include "Utilities/ConstantExpressions.hpp"
+
+namespace {
+// Struct for sorting elements.
+struct SparseMatrixElement {
+  SparseMatrixElement(const size_t row, const size_t column, const double val)
+      : row_dest(row), column_src(column), value(val) {}
+  size_t row_dest, column_src;
+  double value;
+};
+}  // namespace
+
+SparseMatrixFiller::SparseMatrixFiller(const size_t num_cols,
+                                       const bool use_map_method)
+    : num_rows_(num_cols),
+      num_cols_(num_cols),
+      use_map_method_(use_map_method) {
+  if (not use_map_method) {
+    matrix_elements_.assign(square(num_cols), 0.0);
+  }
+}
+
+void SparseMatrixFiller::add(const double element, const size_t dest_index,
+                             const size_t src_index) {
+  if (not use_map_method_) {
+    matrix_elements_[src_index + num_cols_ * dest_index] += element;
+  } else {
+    const std::pair<size_t, size_t> index(dest_index, src_index);
+    const auto iter = element_index_.find(index);
+    if (iter != element_index_.end()) {
```

Comment:
[optional] you can do
```cpp
if (const auto iter = element_index_.find(index); iter != element_index_.end()) {
...
```
for tighter scope. I don't think you can easily do both iter and index? Maybe:
```cpp
if (const auto [index, iter] = std::tuple{std::pair{dest_index, src_index}, element_index_.find(index)};
  iter != element_index_.end()) {
```